app/service/user.js

```diff
@@ -0,0 +1,163 @@
+'use strict';
+const Service = require('egg').Service;
+
+class UserService extends Service {
+    constructor(ctx) {
+        super(ctx);
+    }
+
+    async create({ basicInfo }) {
+        const { ctx } = this;
+        const res = {};
+
+        const userObj = { basicInfo }
+        const result = await ctx.model.User.create(userObj);
+
+        res.data = { id: result._id, createdAt: result.createdAt }
+        res.code = 1
+
+        return res
+    }
+
+    async find({ pageNum = 1, pageSize = 10 }) {
+        const { ctx } = this;
+        const res = {};
+
+        let users = []
+        let count = 0
+        let skip = ((Number(pageNum)) - 1) * Number(pageSize)
+
+        users = await ctx.model.User.find({ valid: true }, 'basicInfo createdAt').skip(skip).limit(Number(pageSize)).sort({ createdAt: -1 }).exec()
+        count = await ctx.model.User.countDocuments({}).exec()
+
+        res.code = 1
+        res.data = { users, count }
+
+        return res
+    }
+
+    async findById({ id }) {
+        const { ctx } = this;
+        const res = {};
+        const result = await ctx.model.User.findById(id);
+        if (result) {
+            res.data = result;
+            res.code = 1;
+        } else {
+            res.data = result;
+            res.code = -1;
+        }
+        return res;
+    }
+
+    async follow({ sourceId, targetId }) {
+        const { ctx } = this;
+        const res = {};
+        const sourceUser = await ctx.model.User.findById(sourceId);
+        if (!sourceUser) {
+            ctx.logger.warn("user.addFollower - user %s doesn't exist", sourceId)
+        }
+
+        const targetUser = await ctx.model.User.findById(targetId);
+        if (!targetUser) {
+            ctx.logger.warn("user.follow - user %s doesn't exist", targetId)
+        }
+
+        if (!sourceUser || !targetUser) {
+            res.code = -1
+            res.msg = "neither source nor target user doesn't exist"
+            return res
+        }
+
+        const sourceActionRes = await ctx.model.User.findByIdAndUpdate(sourceId, { $addToSet: { following: { userId: targetId, userName: targetUser.basicInfo.name } } }, { new: true, select: "_id" });
```

Comment:
that is an anti-pattern design, as once you want to you want to get the followers list, it will be very quick with the name field. 

if  the name changes, take weixin for example, weixin will update it once you view the detail of that follower, here I intent to the same thing.

Reply:
if same follow message come multiple times, can you how $addToSet handle that.